### PR TITLE
OCPRHV-589: Initial oVirt provider support

### PR DESCRIPTION
**Pull request by mmartinv**
# Assisted Pull Request

## Description

Implement the oVirt provider support in assisted service.

- Should this PR be tested by the reviewer? No
- Is this PR relying on CI for an e2e test run? No
- Should this PR be tested in a specific environment? Yes, oVirt must be used to host the VMs
- Any logs, screenshots, etc that can help with the review process? No

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @YuviGold 
/cc @eliorerz
/cc @avishayt 
/cc @eranco74
/cc @michaellevy101
/cc @filanov

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

### CODE REVIEW

File: internal/common/common.go [hunk 1]
```diff
@@ -241,3 +244,45 @@ func GetHostNTPSources(db *gorm.DB, host *models.Host) (string, error) {
...
+		}
+		if block.Type != "CERTIFICATE" || len(block.Headers) != 0 {
+			continue
+		}
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, false
+		}
+		certs = append(certs, *cert)
+	}
```
Comment (Gal-Zaidman) [inline internal/common/common.go:264]: Usage of `break` and `continue` statement are confusing, and together it just looks bad.
How about:
```golang
	for len(pemCerts) > 0 {
		var block *pem.Block
		block, pemCerts = pem.Decode(pemCerts)
		if block == nil {
			break
		}
		if block.Type == "CERTIFICATE" && len(block.Headers) == 0 {
				cert, err := x509.ParseCertificate(block.Bytes)
				if err != nil {
					return nil, false
				}
				certs = append(certs, *cert)
				}
				cert, err := x509.ParseCertificate(block.Bytes)
				if err != nil {
					return nil, false
				}
				certs = append(certs, *cert)
		}
	}
```

File: internal/common/common.go [hunk 2]
```diff
@@ -241,3 +244,45 @@ func GetHostNTPSources(db *gorm.DB, host *models.Host) (string, error) {
...
+	return certs, true
+}
+
+func VerifyCaBundle(pemCerts []byte) error {
```
Comment (Gal-Zaidman) [inline internal/common/common.go:268]: Please add a doc on exported function

File: internal/ignition/ignition.go [hunk 3]
```diff
@@ -388,6 +396,30 @@ func (g *installerGenerator) UploadToS3(ctx context.Context) error {
 	return uploadToS3(ctx, g.workDir, g.cluster, g.s3Client, g.log)
 }
 
+func (g *installerGenerator) CreateOvirtConfig() (string, error) {
```
Comment (Gal-Zaidman) [inline internal/ignition/ignition.go:399]: Please add a doc on exported function

File: internal/ignition/ignition.go [hunk 4]
```diff
@@ -421,6 +453,17 @@ func (g *installerGenerator) Generate(ctx context.Context, installConfig []byte,
...
+	// Create the 'ovirt-config.yaml' file required by the installer for
+	// the oVirt provider and append the OVIRT_CONFIG to the envirionment
+	// variables
+	if g.cluster.Platform.Type == models.PlatformTypeOvirt {
+		ovirtConfigPath, e := g.CreateOvirtConfig()
+		if e != nil {
+			return e
+		}
+		envVars = append(envVars, "OVIRT_CONFIG="+ovirtConfigPath)
+	}
```
Comment (Gal-Zaidman) [inline internal/ignition/ignition.go:465]: This part should be added as a function the provider code. 
All provider related stuff should be in the provider dir

File: internal/provider/ovirt/installConfig.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		platform.URL = fmt.Sprintf("https://%s/ovirt-engine/api", *clusterPlatform.Fqdn)
+		platform.Username = *clusterPlatform.Username
+		platform.Password = *clusterPlatform.Password
+		platform.Insecure = *clusterPlatform.Insecure
+		platform.CaBundle = clusterPlatform.CaBundle
+		platform.ClusterID = *clusterPlatform.ClusterID
+		platform.StorageDomainID = *clusterPlatform.StorageDomainID
+		platform.NetworkName = *clusterPlatform.NetworkName
+		platform.VnicProfileID = clusterPlatform.VnicProfileID
+	}
```
Comment (Gal-Zaidman) [inline internal/provider/ovirt/installConfig.go:23]: Why don't you have the nil stuff like vsphere?
I understood from @eliorerz that it was needed if someone would have change the platform back

File: internal/common/common.go [hunk 6]
```diff
@@ -241,3 +244,45 @@ func GetHostNTPSources(db *gorm.DB, host *models.Host) (string, error) {
 	}
 	return infraEnv.AdditionalNtpSources, nil
 }
+
+func ParsePemCerts(pemCerts []byte) ([]x509.Certificate, bool) {
+	var certs []x509.Certificate
+	for len(pemCerts) > 0 {
```
Comment (Gal-Zaidman) [inline internal/common/common.go:250]: `for len(pemCerts) > 0` ? -- is it decreasing or something?
maybe you meant `if` ?

File: internal/common/common.go [hunk 7]
```diff
@@ -241,3 +244,45 @@ func GetHostNTPSources(db *gorm.DB, host *models.Host) (string, error) {
...
+	}
+	certPool := x509.NewCertPool()
+	for i := range certs {
+		certPool.AddCert(&certs[i])
+	}
+	verifyOptions := x509.VerifyOptions{
+		Roots: certPool,
+	}
+	if _, err := certs[0].Verify(verifyOptions); err != nil {
+		fmt.Printf("Validation error: %e", err)
```
[the remark this answers is not in the document]
Comment (mmartinv) [reply internal/common/common.go:285]: You are right, I will fix it

File: internal/common/common.go (hunk 6 above)
Comment (mmartinv) [reply internal/common/common.go:259]: I updated the function to be more readable

File: swagger.yaml [hunk 8]
```diff
@@ -6046,6 +6046,59 @@ paths:
             $ref: '#/definitions/error'
 
 definitions:
+  ovirt-platform:
+    type: object
+    description: oVirt platform specific configuration upon which to perform the installation
```
Comment (avishayt) [inline swagger.yaml:6051]:
```suggestion
    description: oVirt platform-specific configuration upon which to perform the installation.
```

File: swagger.yaml [hunk 9]
```diff
@@ -6046,6 +6046,59 @@ paths:
...
+    properties:
+      fqdn:
+        type: string
+        description: The oVirt's engine full qualified domain name
```
Comment (avishayt) [inline swagger.yaml:6055]:
```suggestion
        description: The oVirt's engine fully qualified domain name.
```

File: swagger.yaml [hunk 10]
```diff
@@ -6046,6 +6046,59 @@ paths:
...
+      vnic_profile_id:
+        type: string
+        description: The oVirt VNIC profile ID.
+        format: uuid
+        x-nullable: true
+      workers:
+        type: integer
+        default: 0
+        description: Number of additional oVirt workers to be deployed by the Machine API.
+        x-nullable: true
```
Comment (avishayt) [inline swagger.yaml:6099]: I don't think we want to support adding workers during installation in a method other than Assisted.  Why not do it in day2 on the installed cluster?

File: swagger.yaml [hunk 11]
```diff
@@ -5958,6 +5958,55 @@ paths:
...
 definitions:
+  ovirt-platform:
+    type: object
+    description: oVirt platform specific configuration upon which to perform the installation
+    required:
+      - fqdn
+      - username
+      - password
+      - cluster_id
+      - storage_domain_id
```
[the remark this answers is not in the document]
Comment (mmartinv) [reply swagger.yaml]: I modified the swagger specification to look like vsphere's so no field is required now

File: internal/common/common.go [hunk 12]
```diff
@@ -248,3 +251,50 @@ func GetHostNTPSources(db *gorm.DB, host *models.Host) (string, error) {
...
+			if err != nil {
+				return nil, false
+			}
+			certs = append(certs, *cert)
+		}
+	}
+	return certs, true
+}
+
+// Verifies the full CA Chain contained in a byte array.
```
Comment (Gal-Zaidman) [inline internal/common/common.go:276]: The doc text of a function always needs to start with the function name.
 // VerifyCaBundle verifies the full CA Chain contained in a byte array.
....

File: internal/common/common.go [hunk 13]
```diff
@@ -248,3 +251,50 @@ func GetHostNTPSources(db *gorm.DB, host *models.Host) (string, error) {
...
+	}
+	certPool := x509.NewCertPool()
+	for i := range certs {
+		certPool.AddCert(&certs[i])
+	}
+	verifyOptions := x509.VerifyOptions{
+		Roots: certPool,
+	}
+	if _, err := certs[0].Verify(verifyOptions); err != nil {
+		return errors.New("Unable to verify the full CA Chain")
```
Comment (Gal-Zaidman) [inline internal/common/common.go:297]: errors should start with lower case

File: internal/ignition/ignition.go [hunk 14]
```diff
@@ -429,10 +432,20 @@ func (g *installerGenerator) Generate(ctx context.Context, installConfig []byte,
...
+			log.WithError(err).Errorf("Failed to run pre manifests creation hook '%s'", g.cluster.Platform.Type)
+			return err
+		}
 		err = g.runCreateCommand(ctx, installerPath, "manifests", envVars)
 		if err != nil {
 			return err
 		}
+		err = g.providerRegistry.PostCreateManifestsHook(g.cluster.Platform.Type, g.cluster.Platform, &envVars, g.workDir)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to run post manifests creation hook '%s'", g.cluster.Platform.Type)
```
Comment (Gal-Zaidman) [inline internal/ignition/ignition.go:446]: errors should start with lower case

File: internal/provider/ovirt/ignition.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	URL := fmt.Sprintf("https://%s/ovirt-engine/api", *platformParams.Ovirt.Fqdn)
+	oVirtConfig := &OvirtConfig{
+		URL:      URL,
+		Username: swag.StringValue(platformParams.Ovirt.Username),
+		Password: platformParams.Ovirt.Password.String(),
+		Insecure: swag.BoolValue(platformParams.Ovirt.Insecure),
+		CaBundle: swag.StringValue(platformParams.Ovirt.CaBundle),
+	}
+	ovirtConfigPath := filepath.Join(workDir, "ovirt-config.yaml")
```
Comment (Gal-Zaidman) [inline internal/provider/ovirt/ignition.go:33]: where is `workDir`? I wonder if it is safe to write to it.
Anyway write the file with ".ovirt-config.yaml"

File: internal/provider/ovirt/ignition.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+	if platformParams.Ovirt == nil {
+		return "", nil
+	}
+	URL := fmt.Sprintf("https://%s/ovirt-engine/api", *platformParams.Ovirt.Fqdn)
+	oVirtConfig := &OvirtConfig{
+		URL:      URL,
+		Username: swag.StringValue(platformParams.Ovirt.Username),
+		Password: platformParams.Ovirt.Password.String(),
+		Insecure: swag.BoolValue(platformParams.Ovirt.Insecure),
+		CaBundle: swag.StringValue(platformParams.Ovirt.CaBundle),
```
Comment (Gal-Zaidman) [inline internal/provider/ovirt/ignition.go:31]: Any of those params are exposed when you inspect the network calls?

**Review by Gal-Zaidman — COMMENTED** [review]
Code looks mostly good, left some nitpicks.
The biggest concern that I have is about providing the password on swagger, like we talked about in the meeting those params will appear on the API call and we should pass them securely by using https and have a validation that will prevent users from passing does credentials in an unsecured manner

File: internal/common/common.go [hunk 17]
```diff
@@ -248,3 +251,50 @@ func GetHostNTPSources(db *gorm.DB, host *models.Host) (string, error) {
 	}
 	return infraEnv.AdditionalNtpSources, nil
 }
+
+// Returns a list of certificate objects extracted from a byte array
+// The certificates within the byte array must be PEM encoded
+func ParsePemCerts(pemCerts []byte) ([]x509.Certificate, bool) {
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply internal/common/common.go:257]: Consider adding golangci-lint to alert for these errors.

File: internal/common/common.go [hunk 18]
```diff
@@ -248,3 +251,50 @@ func GetHostNTPSources(db *gorm.DB, host *models.Host) (string, error) {
...
+// starting with the leaf CA certificate followed by the
+// subsequents intermediate CAs and ending with the root CA
+func VerifyCaBundle(pemCerts []byte) error {
+	var certs []x509.Certificate
+	var ok bool
+	if certs, ok = ParsePemCerts(pemCerts); !ok {
+		return errors.New("Unable to parse the CA Bundle")
+	}
+	if len(certs) == 0 {
+		return errors.New("The CA bundle does not contain any PEM certificate")
```
Comment (janosdebugs) [inline internal/common/common.go:287]: Errors in Go should start with lower case.

File: internal/provider/ovirt/ignition.go [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/go-openapi/swag"
+	"github.com/openshift/assisted-service/models"
+	"gopkg.in/yaml.v2"
+)
+
+type OvirtConfig struct {
```
Comment (janosdebugs) [inline internal/provider/ovirt/ignition.go:13]: @Gal-Zaidman we should really move this to go-ovirt-client.

File: internal/provider/registry/registry_test.go [hunk 20]
```diff
@@ -25,9 +27,19 @@ var (
 const dummy = "dummy"
 const invalidInventory = "{\"system_vendor\": \"invalid\"}"
 
+const ovirtFqdn = "ovirt.example.com"
+const ovirtInsecure = true
```
Comment (janosdebugs) [inline internal/provider/registry/registry_test.go:31]: Please don't propagate the use of insecure. It's a very bad practice and is copied all over projects. Please use proper certificate verification, even in examples.

File: internal/provider/registry/registry_test.go [hunk 21]
```diff
@@ -352,6 +486,28 @@ func createVspherePlatformParams() *models.Platform {
...
 
+func createOvirtPlatformParams() *models.Platform {
+	Password := strfmt.Password(ovirtPassword)
+	ClusterID := strfmt.UUID(ovirtClusterID)
+	StorageDomainID := strfmt.UUID(ovirtStorageDomainID)
+	VnicProfileID := strfmt.UUID(ovirtVnicProfileID)
+
+	ovirtPlatform := models.OvirtPlatform{
+		Fqdn:            swag.String(ovirtFqdn),
+		Insecure:        swag.Bool(ovirtInsecure),
```
Comment (janosdebugs) [inline internal/provider/registry/registry_test.go:497]: Same, please don't propagate the use of insecure.

**Review by Gal-Zaidman — APPROVED** [review]
Changes looks good to me

File: swagger.yaml [hunk 22]
```diff
@@ -6046,6 +6046,53 @@ paths:
...
+    properties:
+      fqdn:
+        type: string
+        description: The oVirt's engine fully qualified domain name.
+        x-nullable: true
+      username:
+        type: string
+        description: The user name to use to connect to the oVirt instance.
+        x-nullable: true
+      password:
```
Comment (avishayt) [inline swagger.yaml:6061]: In the past, we tried not to return user passwords from the service.  For example, once pull secret is specified via the API, it cannot be retrieved - we have a boolean pull_secret_set to indicate that some value was provided.  I realize that we have the same issue with the vSphere API - I just opened a task to fix that.

File: internal/provider/ovirt/base.go [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+
+// Name returns the name of the provider
+func (p *ovirtProvider) Name() models.PlatformType {
+	return models.PlatformTypeOvirt
+}
+
+func (p *ovirtProvider) IsHostSupported(host *models.Host) (bool, error) {
+	hostInventory, err := common.UnmarshalInventory(host.Inventory)
+	if err != nil {
+		return false, fmt.Errorf("error marshaling host to inventory, error %w", err)
```
Comment (filanov) [inline internal/provider/ovirt/base.go:32]: prefer to use errors.Wrapf

File: pkg/generator/generator.go [hunk 24]
```diff
@@ -91,7 +94,7 @@ func (k *installGenerator) GenerateInstallConfig(ctx context.Context, cluster co
 	if k.Config.DummyIgnition {
 		generator = ignition.NewDummyGenerator(clusterWorkDir, &cluster, k.s3Client, log)
 	} else {
-		generator = ignition.NewGenerator(clusterWorkDir, installerCacheDir, &cluster, releaseImage, k.Config.ReleaseImageMirror, k.Config.ServiceCACertPath, k.Config.InstallInvoker, k.s3Client, log, k.operatorsApi)
+		generator = ignition.NewGenerator(clusterWorkDir, installerCacheDir, &cluster, releaseImage, k.Config.ReleaseImageMirror, k.Config.ServiceCACertPath, k.Config.InstallInvoker, k.s3Client, log, k.operatorsApi, k.providerRegistry)
```
Comment (filanov) [inline pkg/generator/generator.go:97]: please break long lines

**Comment by filanov** [comment]
Overall looks good to me, besides the comment about the password.

File: internal/provider/ovirt/base.go (hunk 23 above)
Comment (Gal-Zaidman) [reply internal/provider/ovirt/base.go:32]: We usually use errors.New on new error we want to trow and fmt.errorf with a %w format on wrapped errors.
See for a comparison:
- https://stackoverflow.com/questions/61933650/whats-the-difference-between-errors-wrapf-errors-errorf-and-fmt-errorf
- https://go.dev/blog/go1.13-errors

Anyway I can go either way, its just the way we did it through the project

File: swagger.yaml (hunk 22 above)
Comment (mmartinv) [reply swagger.yaml:6061]: Would it work for you if I workaround the issue in this PR by overwriting the ovirt password with `nil` just before the API response is delivered? If you agree, I will add a FIXME and open a bug to do the final implementation.

File: internal/provider/ovirt/base.go (hunk 23 above)
Comment (mmartinv) [reply internal/provider/ovirt/base.go:32]: @filanov are you ok with maintaining the current code?

File: swagger.yaml (hunk 22 above)
Comment (avishayt) [reply swagger.yaml:6061]: That would be confusing to users because they set it and it disappears.

Comment (avishayt) [reply swagger.yaml:6061]: I'm OK with handling this in a separate PR

File: internal/provider/ovirt/base.go (hunk 23 above)
Comment (filanov) [reply internal/provider/ovirt/base.go:32]: yes

**Review by avishayt — APPROVED** [review]

**Comment by janosdebugs** [comment]
Congrats! This was a lot of worj.

**Pull request merged**